src/S3Client.php

```diff
@@ -71,6 +86,16 @@ class S3Client {
    * @return \Aws\S3\S3Client
    */
   public static function factory($config = array(), $bucket = NULL) {
+    // Attach doctrine cache for validation here, because this method is always
+    // called to get the $client param of other methods in this class.
+    $stream_wrapper_config = StreamWrapperConfiguration::fromDrupalVariables();
```

Comment:
This is the line that's cause the test to fail.

The way we stub procedural calls like variable_get() is to drop them into a trait that can then be swapped in a sub class. For example, StreamWrapperConfiguration uses Drupal\amazons3\DrupalAdapter\Bootstrap to call variable get, while the StreamWrapperConfiguration stub under the tests directory uses Drupal\amazons3Test\DrupalAdapter\Bootstrap.

In other words, this comes down to this code not injecting the StreamWrapperConfiguration class so we can swap it during tests.

I think there's two followups here to fix this:

1. Let's make a class property with the string of the full class to call here. Then we can add a setter around it to override it in tests, and have the default be \Drupal\amazons3\StreamWrapperConfiguration.
1. This adds some strange coupling conceptually between the S3 client and the stream wrapper configuration class. I was probably a little too narrow in calling the class StreamWrapperConfiguration. Let's copy it to AmazonS3Configuration? Or just Configuration since it's namespaced? Then we mark the StreamWrapperConfiguration class as `@deprecated` and all of it's methods just call out to the Configuration class (preserving backwards compatibility). In fact... I think this refactor could be done as an entirely separate PR. We get that one in first before finishing this one off.

Sound good?